SearchActivity.py

```diff
@@ -187,142 +180,64 @@ def _write_scores_to_clipboard(self, button=None):
-        self._new_tube_common(True)
+        self.after_share_join(True)
 
     def _joined_cb(self, activity):
         """ ...or join an exisiting share. """
         self._new_tube_common(False)
 
     def _new_tube_common(self, sharer):
         """ Joining and sharing are mostly the same... """
-        if self._shared_activity is None:
-            _logger.debug("Error: Failed to share or join activity ... \
-                _shared_activity is null in _shared_cb()")
-            return
+        self.after_share_join(False)
 
-        self.initiating = sharer
+    def after_share_join(self, sharer):
         self.waiting_for_hand = not sharer
-
-        self.conn = self._shared_activity.telepathy_con
-        self.tubes_chan = self._shared_activity.telepathy_tubes_chan
-        self.text_chan = self._shared_activity.telepathy_text_chan
-
-        self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].connect_to_signal(
-            'NewTube', self._new_tube_cb)
-
-        if sharer:
-            _logger.debug('This is my activity: making a tube...')
-            id = self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].OfferDBusTube(
-                SERVICE, {})
-        else:
-            _logger.debug('I am joining an activity: waiting for a tube...')
-            self.tubes_chan[TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].ListTubes(
-                reply_handler=self._list_tubes_reply_cb,
-                error_handler=self._list_tubes_error_cb)
         self._game.set_sharing(True)
 
     def _list_tubes_reply_cb(self, tubes):
         """ Reply to a list request. """
         for tube_info in tubes:
             self._new_tube_cb(*tube_info)
 
-    def _list_tubes_error_cb(self, e):
-        """ Log errors. """
-        _logger.debug('Error: ListTubes() failed: %s' % (e))
-
-    def _new_tube_cb(self, id, initiator, type, service, params, state):
-        """ Create a new tube. """
-        _logger.debug('New tube: ID=%d initator=%d type=%d service=%s \
-params=%r state=%d' % (id, initiator, type, service, params, state))
-
-        if (type == TelepathyGLib.TubeType.DBUS and service == SERVICE):
-            if state == TelepathyGLib.TubeState.LOCAL_PENDING:
-                self.tubes_chan[
-                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES].AcceptDBusTube(id)
-
-            tube_conn = TubeConnection(
-                self.conn, self.tubes_chan[
-                    TelepathyGLib.IFACE_CHANNEL_TYPE_TUBES], id,
-                group_iface=self.text_chan[
-                    TelepathyGLib.IFACE_CHANNEL_INTERFACE_GROUP])
-
-            self.chattube = ChatTube(tube_conn, self.initiating,
-                                     self.event_received_cb)
-
     def _setup_dispatch_table(self):
         ''' Associate tokens with commands. '''
         self._processing_methods = {
             'n': [self._receive_new_game, 'get a new game grid'],
             'p': [self._receive_dot_click, 'get a dot click'],
```

Comment:
It'll be great to have more informative key names for `n` and `p` like `new_game` and `played`.